subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java

```diff
@@ -122,78 +106,53 @@ public void triggerWatchProbe(String path) {
-            .filter(oldWatchedHierarchy -> !watchedHierarchies.contains(oldWatchedHierarchy))
-            .collect(Collectors.toCollection(() -> new ArrayList<>(oldWatchedHierarchies.size())));
-        List<File> hierarchiesToStartWatching = watchedHierarchies.stream()
-            .filter(newWatchedHierarchy -> !oldWatchedHierarchies.contains(newWatchedHierarchy))
-            .collect(Collectors.toCollection(() -> new ArrayList<>(watchedHierarchies.size())));
-
         oldProbedHierarchies.stream()
             .filter(oldProbedHierarchy -> !probedHierarchies.contains(oldProbedHierarchy))
             .forEach(probedHierarchy -> {
-                stopWatchingProbeForHierarchy(probedHierarchy);
-                probeRegistry.disarmWatchProbe(probedHierarchy);
+                File probeDirectory = probeRegistry.getProbeDirectory(probedHierarchy);
+                disarmWatchProbeForHierarchy(probedHierarchy, probeDirectory);
             });
 
-        if (!hierarchiesToStopWatching.isEmpty()) {
-            stopWatchingHierarchies(hierarchiesToStopWatching);
-        }
-        if (!hierarchiesToStartWatching.isEmpty()) {
-            hierarchiesToStartWatching.forEach(locationToWatchValidator::validateLocationToWatch);
-            startWatchingHierarchies(hierarchiesToStartWatching);
-        }
-
         probedHierarchies.stream()
             .filter(newProbedHierarchy -> !oldProbedHierarchies.contains(newProbedHierarchy))
             .forEach(probedHierarchy -> {
-                startWatchingProbeForHierarchy(probedHierarchy);
-                probeRegistry.armWatchProbe(probedHierarchy);
+                File probeDirectory = probeRegistry.getProbeDirectory(probedHierarchy);
+                // Make sure the directory exists, this can be necessary when
+                // included builds are evaluated with configuration cache
+                //noinspection ResultOfMethodCallIgnored
+                probeDirectory.mkdirs();
+                armWatchProbeForHierarchy(probedHierarchy, probeDirectory);
             });
-
-        LOGGER.info("Watching {} directory hierarchies to track changes", watchedHierarchies.size());
     }
 
-    protected abstract void startWatchingHierarchies(Collection<File> hierarchiesToWatch);
-
-    protected abstract void startWatchingProbeForHierarchy(File hierarchyToWatch);
-
-    protected abstract void stopWatchingHierarchies(Collection<File> hierarchiesToWatch);
-
-    protected abstract void stopWatchingProbeForHierarchy(File hierarchyToWatch);
+    protected FileHierarchySet updateWatchedHierarchies(SnapshotHierarchy root) {
+        return resolveWatchedFiles(watchableHierarchies, root);
+    }
 
-    public interface FileSystemLocationToWatchValidator {
-        FileSystemLocationToWatchValidator NO_VALIDATION = location -> {
-        };
+    @OverridingMethodsMustInvokeSuper
+    protected void armWatchProbeForHierarchy(File probedHierarchy, File probeDirectory) {
```

Comment:
I'd prefer using template method instead of `@OverridingMethodsMustInvokeSuper`. That adds an empty implementation in `HierarchicalWatcherUpdater`, though it makes it clearer what happens in what order.

Reply:
While I share your dislike of overrides in general, this one I'm quite okay with. My argument here is that arming a watch probe involves watching the directory for the non-hierarchical case.